### PR TITLE
add GMRES restart option

### PR DESCRIPTION
**Pull request by joergbuchwald**
add restart option for the GMRES solver (Eigen-unsupported). 
It sets the number of iterations before a restart is conducted. It can improve the convergence behavior. However, increasing the value comes with a tradeoff in time and memory. First tests suggest that the default value (30) is already quite optimal in most cases.

1. [x] Feature description was added to the [changelog](https://github.com/ufz/ogs/wiki/Release-notes-6.3.1)
2. [ ] Tests covering your feature were added?
3. [x] Any new feature or behavior change was documented?

### CODE REVIEW

File: MathLib/LinAlg/Eigen/EigenOption.h [hunk 1]
```diff
@@ -47,6 +47,7 @@ struct EigenOption final
 #ifdef USE_EIGEN_UNSUPPORTED
     /// Scaling the coefficient matrix and the RHS bector
     bool scaling;
+    int restart;
```
Comment (endJunction) [inline MathLib/LinAlg/Eigen/EigenOption.h:50]: Only used by GMRES, isn't it? If so, add a comment.

File: MathLib/LinAlg/Eigen/EigenLinearSolver.cpp [hunk 2]
```diff
@@ -265,6 +293,18 @@ void EigenLinearSolver::setOption(BaseLib::ConfigTree const& option)
             "scaling is not available.");
 #endif
     }
+    if (auto restart =
+            //! \ogs_file_param{prj__linear_solvers__linear_solver__eigen__restart}
```
Comment (endJunction) [inline MathLib/LinAlg/Eigen/EigenLinearSolver.cpp:297]: tag documentation is required. Preferably refer to the documentation of the `EigenOption::restart` member.

File: MathLib/LinAlg/Eigen/EigenLinearSolver.cpp [hunk 3]
```diff
@@ -119,8 +120,35 @@ class EigenIterativeLinearSolver final : public EigenLinearSolverBase
...
+    _solver.set_restart(restart);
+    INFO("-> set restart value: {:d}", _solver.get_restart());
+}
+
+template <>
+void EigenIterativeLinearSolver<Eigen::GMRES<EigenMatrix::RawMatrixType, Eigen::IncompleteLUT<double>>>::setRestart(int const restart)
+{
+    _solver.set_restart(restart);
+    INFO("-> set restart value: {:d}", _solver.get_restart());
+}
```
Comment (TomFischer) [inline MathLib/LinAlg/Eigen/EigenLinearSolver.cpp:150]: Why is the setting of the `restart` value only possible for `IdentityPreconditioner`, `DiagonalPreconditioner`, and `IncompleteLUT`? The restart should be independent from the preconditioner at all.

Comment (endJunction) [reply MathLib/LinAlg/Eigen/EigenLinearSolver.cpp:150]: Not sure if the partial specialization will work here.
@joergbuchwald Please try it.

Comment (joergbuchwald) [reply MathLib/LinAlg/Eigen/EigenLinearSolver.cpp:150]: That's true. The problem I had, was with the T_SOLVER type. It is not instantiated via a general preconditioner type but via the specific preconditioners. Therefore, I had to make the code specialization for each preconditioner.

Comment (endJunction) [reply MathLib/LinAlg/Eigen/EigenLinearSolver.cpp:150]: Something like:
```c++
template <typename Preconditioner>
￼void EigenIterativeLinearSolver<Eigen::GMRES<EigenMatrix::RawMatrixType, Preconditioner>>::setRestart(int const restart)
```

File: MathLib/LinAlg/Eigen/EigenLinearSolver.cpp [hunk 4]
```diff
@@ -119,8 +120,36 @@ class EigenIterativeLinearSolver final : public EigenLinearSolverBase
...
+void EigenIterativeLinearSolver<Eigen::GMRES<
+    EigenMatrix::RawMatrixType,
+    Eigen::DiagonalPreconditioner<double>>>::setRestart(int const restart)
+{
+    _solver.set_restart(restart);
+    INFO("-> set restart value: {:d}", _solver.get_restart());
+}
+
+template <>
+void EigenIterativeLinearSolver<Eigen::GMRES<EigenMatrix::RawMatrixType, Eigen::IncompleteLUT<double>>>::setRestart(int const restart)
```
Comment (endJunction) [inline MathLib/LinAlg/Eigen/EigenLinearSolver.cpp:147]: @joergbuchwald clang-format. Use pre-commit checks.

**Comment by endJunction** [comment]
@joergbuchwald Small testcase using the option, something from SteadyStateDiffusion, for example, would be good.

**Pull request merged**